### PR TITLE
Hyundai: Car Port for Nexo 2021

### PR DESCRIPTION
**Pull request by sunnyhaibin**
**Checklist**
- [x] added entry to CarInfo in selfdrive/car/*/values.py and ran `selfdrive/car/docs.py` to generate new docs
- [x] test route added to [routes.py](https://github.com/commaai/openpilot/blob/master/selfdrive/car/tests/routes.py)
- [x] route with stock system: `a74afe0cf708748f|2023-09-02--19-56-27`
- [x] route with openpilot: `a74afe0cf708748f|2023-09-10--11-18-49`
- [x] harness Type: [Hyundai H](https://github.com/commaai/openpilot/wiki/Hyundai-Kia-Genesis#harness-guide)
- [x] architecture: CAN

**Link to all physical parts during installation**
  - [comma 3X](https://comma.ai/shop/comma-3x)
  - [car harness kit](https://comma.ai/shop/car-harness)
    - [Hyundai H](https://github.com/commaai/openpilot/wiki/Hyundai-Kia-Genesis#harness-guide)

**Requires**
- https://github.com/commaai/panda/pull/1637
- https://github.com/commaai/opendbc/pull/933
- https://github.com/commaai/openpilot/pull/29772

Thanks to community 2021 Hyundai Nexo owner SimonYang402.

### CODE REVIEW

**Comment by sunnyhaibin** [comment]
@sshane This port is ready to review.

File: selfdrive/car/hyundai/carstate.py [hunk 1]
```diff
@@ -120,11 +122,12 @@ def update(self, cp, cp_cam):
...
-      if self.CP.carFingerprint in HYBRID_CAR:
-        ret.gas = cp.vl["E_EMS11"]["CR_Vcu_AccPedDep_Pos"] / 254.
-      else:
-        ret.gas = cp.vl["E_EMS11"]["Accel_Pedal_Pos"] / 254.
+    if self.CP.carFingerprint in (EV_CAR | HYBRID_CAR | FCEV_CAR):
+      gas_msg = "ACCELERATOR" if self.CP.carFingerprint in FCEV_CAR else "E_EMS11"
+      gas_sig = "Accel_Pedal_Pos" if self.CP.carFingerprint in EV_CAR else \
+                "ACCELERATOR_PEDAL" if self.CP.carFingerprint in FCEV_CAR else \
+                "CR_Vcu_AccPedDep_Pos"
+      ret.gas = cp.vl[gas_msg][gas_sig] / 254.
```
Comment (sshane) [inline selfdrive/car/hyundai/carstate.py:130]:
```suggestion
    if self.CP.carFingerprint in (HYBRID_CAR | EV_CAR | FCEV_CAR):
      if self.CP.carFingerprint in FCEV_CAR:
        ret.gas = cp.vl["ACCELERATOR"]["ACCELERATOR_PEDAL"] / 254.
      elif self.CP.carFingerprint in HYBRID_CAR:
        ret.gas = cp.vl["E_EMS11"]["CR_Vcu_AccPedDep_Pos"] / 254.
      else:
        ret.gas = cp.vl["E_EMS11"]["Accel_Pedal_Pos"] / 254.
```

I'd argue this is much easier to read

Comment (sshane) [inline selfdrive/car/hyundai/carstate.py:130]:
```suggestion
    if self.CP.carFingerprint in (HYBRID_CAR | EV_CAR | FCEV_CAR):
      if self.CP.carFingerprint in FCEV_CAR:
        ret.gas = cp.vl["ACCELERATOR"]["ACCELERATOR_PEDAL"] / 254.
      elif self.CP.carFingerprint in HYBRID_CAR:
        ret.gas = cp.vl["E_EMS11"]["CR_Vcu_AccPedDep_Pos"] / 254.
      else:
        ret.gas = cp.vl["E_EMS11"]["Accel_Pedal_Pos"] / 254.
```

I'd argue this is much easier to read

File: selfdrive/car/hyundai/interface.py [hunk 2]
```diff
@@ -145,6 +145,10 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
       ret.mass = 1690.  # from https://www.hyundai-motor.com.tw/clicktobuy/custin#spec_0
       ret.wheelbase = 3.055
       ret.steerRatio = 17.0  # from learner
+    elif candidate == CAR.NEXO_1ST_GEN:
+      ret.mass = 3990 * CV.LB_TO_KG  # https://www.hyundainews.com/assets/documents/original/42768-2021NEXOProductGuideSpecs.pdf
+      ret.wheelbase = 2.79  # https://www.hyundainews.com/assets/documents/original/42768-2021NEXOProductGuideSpecs.pdf
+      ret.steerRatio = 14.19  # https://www.hyundainews.com/assets/documents/original/42768-2021NEXOProductGuideSpecs.pdf
```
Comment (sshane) [inline selfdrive/car/hyundai/interface.py:151]:
```suggestion
      ret.mass = 3990 * CV.LB_TO_KG  # https://www.hyundainews.com/assets/documents/original/42768-2021NEXOProductGuideSpecs.pdf
      ret.wheelbase = 2.79
      ret.steerRatio = 14.19
```

File: selfdrive/car/hyundai/values.py [hunk 3]
```diff
@@ -2057,6 +2073,7 @@ def match_fw_to_car_fuzzy(live_fw_versions) -> Set[str]:
   # which message has the gear. hybrid and EV use ELECT_GEAR
   "use_cluster_gears": {CAR.ELANTRA, CAR.KONA},
   "use_tcu_gears": {CAR.KIA_OPTIMA_G4, CAR.KIA_OPTIMA_G4_FL, CAR.SONATA_LF, CAR.VELOSTER, CAR.TUCSON},
+  "use_ems_gears": {CAR.NEXO_1ST_GEN, },
```
Comment (sshane) [inline selfdrive/car/hyundai/values.py:2076]: Just use `FCEV_CAR`, trying to move away from many hard-coded lists

Comment (sunnyhaibin) [reply selfdrive/car/hyundai/values.py:2136]: Done.

**Comment by sunnyhaibin** [comment]
@sshane can we mark this with `ignore-stale` tag?

**Comment by adeebshihadeh** [comment]
`car port` gets ignored by the stale bot now.

**Pull request closed without merging**